flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -115,577 +144,783 @@ public void testInitStateMap() {
 					if (random.nextBoolean()) {
-						stateMap.remove(key, namespace);
-					} else {
-						assertEquals(keyEntry.getValue(), stateMap.removeAndGetOld(key, namespace));
+						long namespace = namespaceEntry.getKey();
+						removedNamespaces.add(namespace);
+						for (Map.Entry<Integer, String> keyEntry : namespaceEntry.getValue().entrySet()) {
+							int key = keyEntry.getKey();
+							if (random.nextBoolean()) {
+								stateMap.remove(key, namespace);
+							} else {
+								assertEquals(keyEntry.getValue(), stateMap.removeAndGetOld(key, namespace));
+							}
+							totalSize--;
+							if (withSnapshot) {
+								// logical remove with copy-on-write
+								totalSpaceSize++;
+							} else {
+								// physical remove
+								totalSizeIncludingLogicalRemove--;
+								totalSpaceSize -= 2;
+							}
+						}
 					}
-					totalSize--;
-					assertEquals(totalSize, stateMap.size());
-					assertEquals(totalSize, stateMap.totalSize());
 				}
-			}
-		}
-
-		for (long namespace : removedNamespaces) {
-			referenceStates.remove(namespace);
-			assertEquals(0, stateMap.sizeOfNamespace(namespace));
-			assertFalse(stateMap.getKeys(namespace).iterator().hasNext());
-		}
 
-		assertEquals(totalSize * 2, spaceAllocator.getTotalSpaceNumber());
-		verifyState(referenceStates, stateMap);
+				for (long namespace : removedNamespaces) {
+					referenceStates.remove(namespace);
+					// verify namespace related stuff.
+					assertEquals(0, stateMap.sizeOfNamespace(namespace));
+					assertFalse(stateMap.getKeys(namespace).iterator().hasNext());
+				}
+				if (withSnapshot) {
+					snapshot.release();
+				}
+				return new Tuple3<>(totalSize, totalSizeIncludingLogicalRemove, totalSpaceSize);
+			}
+		);
+	}
 
+	/**
+	 * Test close operation.
+	 */
+	@Test
+	public void testClose() {
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap = createStateMapForTesting();
+		putStates(stateMap, null);
 		stateMap.close();
+		assertTrue(stateMap.isClosed());
 		assertEquals(0, stateMap.size());
 		assertEquals(0, stateMap.totalSize());
-		// all spaces should be free
 		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
-		assertTrue(stateMap.isClosed());
 	}
 
 	/**
-	 *  Tests copy-on-write contracts.
+	 * Test with the given function.
+	 *
+	 * @param function a {@link TriFunction} with [totalSizeBeforeFunction, stateMap, referenceStates] as input
+	 *                 parameters and returns the [totalSize, totalSizeIncludingLogicalRemovedKey, totalSpaceSize]
+	 *                 tuple after applying the function.
 	 */
-	@SuppressWarnings("unchecked")
-	@Test
-	public void testCopyOnWriteContracts() throws IOException {
-		TypeSerializer<Integer> keySerializer = IntSerializer.INSTANCE;
-		TypeSerializer<Long> namespaceSerializer = LongSerializer.INSTANCE;
-		TypeSerializer<String> stateSerializer = StringSerializer.INSTANCE;
+	private void testWithFunction(
```

Comment:
The name is not very descriptive. Do we really need this to write our tests? We could also write a test like

```
@Test
public void test() {
  stateMap = createTestStateMap();
  actOnStateMap;
  assertions;
}
```

instead of passing a hard to parse `TriFunction` to some other methods where it is not very clear what it does.

Reply:
The same question here, that when to use functions to reduce duplicated codes and when not? A little bit lost.